src/transformers/models/cogvlm/modeling_cogvlm.py

```diff
@@ -0,0 +1,1132 @@
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+class CogvlmPatchEmbedding(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+
+        self.num_patches = (config.image_size // config.patch_size) ** 2 + 1
+        self.proj = nn.Conv2d(
+            config.num_channels, config.hidden_size, kernel_size=config.patch_size, stride=config.patch_size
+        )
+        self.cls_embedding = nn.Parameter(torch.zeros(1, config.hidden_size))
+        self.position_embedding = nn.Embedding(self.num_patches, config.hidden_size)
+
+    def forward(self, pixel_values: torch.FloatTensor) -> torch.FloatTensor:
+        embeddings = self.proj(pixel_values)
+        embeddings = embeddings.flatten(2).transpose(1, 2)
+        cls_token = self.cls_embedding.expand(embeddings.shape[0], -1, -1)
+        embeddings = torch.cat((cls_token, embeddings), dim=1)
+        embeddings += self.position_embedding.weight.unsqueeze(0)
+        return embeddings
+
+
+class CogvlmVisionAttention(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.num_heads = config.num_attention_heads
+        head_dim = config.hidden_size // config.num_attention_heads
+        self.scale = 1.0 / head_dim**0.5
+        self.query_key_value = nn.Linear(config.hidden_size, config.hidden_size * 3)
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+        self.output_dropout = torch.nn.Dropout(config.dropout_prob)
+
+    def forward(self, hidden_state: torch.FloatTensor) -> torch.FloatTensor:
+        batch_size, sequence_length, _ = hidden_state.shape
+        qkv = self.query_key_value(hidden_state)
+
+        # reshape to (3, batch_size, sequence_length, num_heads, head_dim)
+        qkv = qkv.reshape(batch_size, sequence_length, 3, self.num_heads, -1).permute(2, 0, 1, 3, 4)
+        queries, keys, values = qkv[0], qkv[1], qkv[2]
+
+        import xformers.ops as xops
+
+        out = xops.memory_efficient_attention(
+            queries,
+            keys,
+            values,
+            scale=self.scale,
+        )
+
+        output = self.dense(out.view(batch_size, sequence_length, -1))
+        output = self.output_dropout(output)
+        return output
+
+        # thanks to https://github.com/facebookresearch/xformers/issues/922#issuecomment-1808329588
+        # queries = queries * self.scale
+        # queries = queries.transpose(1, 2)
+        # keys = keys.transpose(1, 2)
+        # values = values.transpose(1, 2)
+        # attention_scores = queries @ keys.transpose(-2, -1)
+
+        # attention_probs = attention_scores.softmax(-1)
+        # attention_output = attention_probs @ values
+        # attention_output = attention_output.transpose(1, 2)
+
+        # output = self.dense(attention_output.reshape(batch_size, sequence_length, -1))
+        # output = self.output_dropout(output)
+        # return output
```

Comment:
Tagging @fxmarty here since he probably knows this best - I'm attempting to convert this `xformers. memory_efficient_attention` by native PyTorch. As suggested in [this issue](https://github.com/facebookresearch/xformers/issues/922#issuecomment-1808329588), as well as [in their docs](https://facebookresearch.github.io/xformers/components/ops.html#module-xformers.ops), this is the code that can be used for that. However, when using this code, the logits don't match, not even at atol=1e-1.

Do you have an idea of could be going on?